File: leet_0005_longest_palindrom_substr___brute.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    n = s.length
    if (n == 1)
        return s
    head = 0
    tail = 1
    i = 0
    while (++i < n){
        L = i - 1
        R = i
        while (R < n && L > -1 && s[L] == s[R]){
            dist = R - L + 1
            if (tail < dist){
                tail = dist
                head = L
            }
            L -= 1
            R += 1
        }
        L = i - 1
        R = i + 1
        while (R < n && L > -1 && s[L] == s[R]){
            dist = R - L + 1
            if (tail < dist){
                tail = dist
                head = L
            }
            L -= 1
            R += 1
        }
    }
    /* if (head == tail)
    {
        if (head * 2 == n)
            return s.substring(head)
        return s[0]
    }*/
    return s.substring(head, head + tail)
}
/*
"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabcaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
"babad"
"cbbd"
"a"
"ac"
"acc"
*/

File: leet_0005_longest_palindrom_substr___brute.go
func longestPalindrome(s string) string {

    size := len(s)
    if size == 1 {
        return s
    }

    head := 0
    tail := 1
    var L, R int
    var dist int
    i := 1

    for i < size {

        L = i - 1
        R = i

        for R < size && L > -1 && s[L] == s[R] {
            dist = R - L + 1
            if tail < dist {
                tail = dist
                head = L
            }
            L--
            R++
        }

        L = i - 1
        R = i + 1

        for R < size && L > -1 && s[L] == s[R] {
            dist = R - L + 1
            if tail < dist {
                tail = dist
                head = L
            }
            L--
            R++
        }
        i++
    }

    return s[head : head + tail]
}
